src/confluent_kafka/schema_registry/schema_registry_client.py

```diff
@@ -671,19 +671,22 @@ class Schema(object):
     Args:
         schema_str (str): String representation of the schema.
 
+        schema_type (str): The schema type: AVRO, PROTOBUF or JSON.
+
         references ([SchemaReference]): SchemaReferences used in this schema.
 
-        schema_type (str): The schema type: AVRO, PROTOBUF or JSON.
+        named_schemas (dict): Named schemas
 
     """
-    __slots__ = ['schema_str', 'references', 'schema_type', '_hash']
+    __slots__ = ['schema_str', 'schema_type', 'references', 'named_schemas', '_hash']
 
-    def __init__(self, schema_str, schema_type, references=[]):
+    def __init__(self, schema_str, schema_type, references=[], named_schemas={}):
         super(Schema, self).__init__()
 
         self.schema_str = schema_str
         self.schema_type = schema_type
         self.references = references
+        self.named_schemas = named_schemas
```

Comment:
@slominskir Any reason why do the user needs to provide `named_schemas`. Can't we fetch it from schema registry like you've done [here](https://github.com/confluentinc/confluent-kafka-python/blob/c973f3ca2a91fd730b511cc153de58a41ae0a965/src/confluent_kafka/schema_registry/avro.py#L385-L395)? This seems to be leaking implementation detail of the library which can be avoided.

Reply:
If I'm remembering correctly in some cases the user may already have the `named_schemas` object so it would be inefficient to require the serde to look it up.   Mostly though, it's probably because I was keeping changes to a minimum and trying to make it possible to use schema references using the odd mix of existing APIs.

For the serializer - you'd need to update code to iterate over the references array and populate named_schemas, probably here: https://github.com/confluentinc/confluent-kafka-python/blob/c973f3ca2a91fd730b511cc153de58a41ae0a965/src/confluent_kafka/schema_registry/avro.py#L206-L207

However, the serializer does not have schema strings for schema references and in the case of it not existing in the registry yet you can't look it up either.  Perhaps require references exist in the registry before a parent schema can be registered (users would need to manually loop over references themselves registering them in order).  Else it could be a separate array added to the Schema object that contains reference schema strings in the same index that corresponds to the SchemaReference array.

Depending on the approach above, the auto register code may need to be updated to loop over references and register them.  This would require the reference schema strings, probably here: https://github.com/confluentinc/confluent-kafka-python/blob/c973f3ca2a91fd730b511cc153de58a41ae0a965/src/confluent_kafka/schema_registry/avro.py#L259-L264

For the deserializer - In many cases the reader and writer schema are the same so if you do move this behavior into the serde you can probably check if reader and writer schema_id match and just use the one lookup that's already being done.

Reply:
Thanks for the context. We are working on to include the schema reference support. There are some more changes needed which I'll push in this PR itself.